src/eigen/lobpcg_hyper_impl.jl

```diff
@@ -43,20 +43,96 @@
+Build a BlockVector containing the given arrays, from left to right.
+This function will fail (for now) if:
+    -the arrays do not all have the same "height" (ie size[1] must match).
+"""
+function make_block_vector(arrays::AbstractArray...)
+    length(arrays) ==0 && error("Empty BlockVector is not currently implemented")
+    n_ref= size(arrays[1])[1]
+    m=0
+    for array in arrays
+        n_i, m_i = size(array)
+        n_ref != n_i && error("The given arrays do not have matching 'height': cannot build a BlockVector out of them.")
+        m += m_i
+    end
+    BlockVector(arrays, (n_ref,m))
+end
+
+
+"""
+Given A and B as two BlockVectors [A1, A2, A3], [B1, B2, B3] form the matrix
+A'B (which is not a BlockVector). block_overlap also has compatible versions with two Arrays. 
+block_overlap always compute some form of adjoint, ie the product A'*B.
+"""
+@views function block_overlap(A::BlockVector, B::BlockVector)
+    rows = A.size[2]
+    cols = B.size[2]
+    ret = similar(A.blocks[1], rows, cols)
+
+    orow = 0  # row offset
+    for (iA, blA) in enumerate(A.blocks)
+        ocol = 0  # column offset
+        for (iB, blB) in enumerate(B.blocks)
+            ret[orow .+ (1:size(blA, 2)), ocol .+ (1:size(blB, 2))] = blA' * blB
+            ocol += size(blB, 2)
+        end
+        orow += size(blA, 2)
+    end
+    ret
+end
+
+block_overlap(blocksA::BlockVector, B) = block_overlap(blocksA, make_block_vector(B))
+block_overlap(A, B) = A' * B #Default fallback method. Note the adjoint.
+
+"""Given A as a BlockVector [A1, A2, A3] this forms the matrix-matrix product
+A * B avoiding a concatenation of the blocks to a dense array. block_mul has compatible versions with two Arrays.
+block_overlap always compute  product A*B (no adjoint).
+"""
+@views function block_mul(Ablock::BlockVector, B)
+    res = Ablock.blocks[1] * B[1:size(Ablock.blocks[1], 2), :]  # First multiplication
+    offset = size(Ablock.blocks[1], 2)
+    for block in Ablock.blocks[2:end]
+        mul!(res, block, B[offset .+ (1:size(block, 2)), :], 1, 1)
+        offset += size(block, 2)
+    end
+    res
 end
 
+block_mul(A, Bblock::BlockVector) = error("Not implemented")
+block_mul(A::Tuple, B::Tuple) = error("not implemented")
+block_mul(A, B) = A * B #Default fallback method.
+
 # Perform a Rayleigh-Ritz for the N first eigenvectors.
-@timing function rayleigh_ritz(X, AX, N)
-    F = eigen(Hermitian(array_mul(X', AX)))
+@timing function rayleigh_ritz(X::BlockVector, AX::BlockVector, N)
+    rayleigh_ritz(block_overlap(X, AX), N) #block_overlap(X,AX) is an AbstractArray, not a BlockVector
+end
+
+@timing function rayleigh_ritz(XAX::AbstractArray, N)
+    F = eigen(Hermitian(XAX))
     F.vectors[:,1:N], F.values[1:N]
 end
 
+@timing function rayleigh_ritz(XAX::CuArray, N)
+    if eltype(XAX) <: Complex
+        vals, vects = CUDA.CUSOLVER.heevd!('V','U',XAX)
+    else
+        vals, vects = CUDA.CUSOLVER.syevd!('V','U',XAX)
+    end
+    vects[:,1:N], vals[1:N]
+end
```

Comment:
Is this really needed? Did you open an issue over at CUDA.jl?

Reply:
Also for this you can just type pirate it (at least for now...) and add it to workarounds/